api/jobs/topEvents.js

```diff
@@ -0,0 +1,98 @@
+const job = require("../parts/jobs/job.js");
+const async = require("async");
+const _ = require("underscore");
+const crypto = require("crypto");
+const common = require("../utils/common.js");
+const countlyApi = {
+    data: {
+        fetch: require("../parts/data/fetch.js"),
+        lib: require("../lib/countly.model.js")
+    }
+};
+const countlyEvents = countlyApi.data.lib.load("event");
+
+/** Class for job of top events widget **/
+class TopEventsJob extends job.Job {
+    /**
+     * Run the job
+     * @param {Db} db connection
+     * @param {done} done callback
+     */
+    run(db, done) {
+        common.db.collection("apps").find({}, {}).toArray(function(errorApps, resultApps) {
+            async.eachSeries(resultApps, function(givenApps, eachSeriesCallback) {
+                const app_id = givenApps._id;
+                const appTimezone = givenApps.timezone;
+                common.db.collection("events").findOne({_id: app_id}, function(errorEvents, result) {
+                    if (!errorEvents && result && Object.prototype.hasOwnProperty.call(result, "list")) {
+                        result.list = _.filter(result.list, function(l) {
+                            return l.indexOf("[CLY]") !== 0;
+                        });
+                        let eventMap = [];
+                        for (let index = 0; index < result.list.length; index++) {
+                            if (result.map) {
+                                let mapKey = result.list[index];
+                                if (typeof result.map[mapKey] === "undefined") {
+                                    result.map[mapKey] = true;
+                                }
+                                if (typeof result.map[mapKey] !== "object" && result.map[mapKey]) {
+                                    eventMap.push(result.list[index]);
+                                }
+                            }
+                            else {
+                                eventMap.push(result.list[index]);
+                            }
+                        }
+                        if (Array.isArray(eventMap)) {
+                            const arrPeriod = ["30days", "hour"];
+                            const appID = String(app_id.toString());
+                            const collectionName = "top_events" + crypto.createHash("sha1").update(appID).digest("hex");
```

Comment:
Since collection will only store couple of documents per app, rather than creating separate collection for each app, let's better create single collection top_events for whole server and use it for all apps, by providing app_id as one of the fields